gh-82012: Deprecate bitwise inversion (~) of bool

timhoffm opened the pull request:
The bitwise inversion operator on bool returns the bitwise inversion of the underlying int value; i.e. `~True == -2` such that `bool(~True) == True`.

It's a common pitfall that users mistake `~` as negation operator and actually want `not`. Supporting `~` is an artifact of bool inheriting from int. Since there is no real use-case for the current behavior, let's deprecate `~` on bool and later raise an error. This removes a potential source errors for users.

Full reasoning: https://github.com/python/cpython/issues/82012#issuecomment-1258705971


<!-- gh-issue-number: gh-82012 -->
* Issue: gh-82012
<!-- /gh-issue-number -->

timhoffm committed:
gh-82012: Deprecate bitwise inversion (~) of bool

The bitwise inversion operator on bool returns the bitwise inversion of the
underlying int value; i.e. `~True == -2` such that `bool(~True) == True`.

It's a common pitfall that users mistake `~` as negation operator and actually
want `not`. Supporting `~` is an artifact of bool inheriting from int. Since there
is no real use-case for the current behavior, let's deprecate `~` on bool and
later raise an error. This removes a potential source errors for users.

Full reasoning: https://github.com/python/cpython/issues/82012#issuecomment-1258705971

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

timhoffm commented on Lib/test/test_bool.py:
```diff
@@ -58,8 +58,10 @@ def test_math(self):
-        self.assertEqual(~False, -1)
-        self.assertEqual(~True, -2)
+        with self.assertWarns(DeprecationWarning):
+            self.assertEqual(~False, -1)
```
The deprecation warning is not issued during testing - and thus the test suite fails.

Can anybody give a hint on why that's the case? When compiling locally, I get the warning shown in the interpreter.

Eclips4 replied:
How I understand, this is a some sort of optimization(?).
Just run this code:
```python
def foo():
    return ~False
```
As you see, there's no call of function `foo`. But you'll see a `DeprecationWarning`.
I doesn't completely understand why is this so, but, simplest way to force this test work looks like that:
```python
with self.assertWarns(DeprecationWarning):
    eval("~False")
```

hauntsaninja replied:
Yeah, the reason is constant folding in the compiler:
```
In [1]: import dis

In [2]: def foo():
   ...:     x = 1 + 2 + 3
   ...:     y = ~False
   ...: 

In [3]: dis.dis(foo)
  2           0 LOAD_CONST               1 (6)
              2 STORE_FAST               0 (x)

  3           4 LOAD_CONST               2 (-1)
              6 STORE_FAST               1 (y)
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE
```

timhoffm replied:
Thanks for the clarification.

I've switched from
```
eval("~False")
```
to
``` 
false = False
~false
```
which also prevents the constant folding and is hopefully less obscure than `eval()`

gvanrossum reviewed (approved):
LGTM. Let's just merge this.

JelleZijlstra commented on Lib/test/test_bool.py:
```diff
@@ -58,8 +58,16 @@ def test_math(self):
-        self.assertEqual(~True, -2)
+        with self.assertWarns(DeprecationWarning):
+            # We need to put the bool in a variable, because the constant
+            # ~False is evaluated at compile time due to constant folding;
```
We should test this behavior separately (doing something like `with assertWarns(DeprecationWarning): exec("~False")`).

timhoffm replied:
Do you mean instead of or in addition to the current test? We originally had `eval("~True")` but I figured it's clearer to have the warning context only around the operation and not around a comparably complex `eval()` or `exec()`. I'm happy to change if there's a benefit of these though.

JelleZijlstra replied:
We should have both. The existing tests check that the deprecation warning is emitted correctly at runtime. The new tests would ensure that if the operation occurs at compile time, we still emit the DeprecationWarning.

JelleZijlstra commented on Doc/whatsnew/3.12.rst:
```diff
@@ -692,6 +692,11 @@ Deprecated
+* The bitwise inversion operator (~) on bool is deprecated. It will throw an
+  error in Python 3.14. Use ``not`` for logical negation of bools instead.
+  In the rare case that you really need the bitwise inversion of the underlying
+  ``int``, convert to int explicitly ``~int(x)``. (Contributed by Tim Hoffmann
```
```suggestion
* The bitwise inversion operator (``~``) on bool is deprecated. It will throw an
  error in Python 3.14. Use ``not`` for logical negation of bools instead.
  In the rare case that you really need the bitwise inversion of the underlying
  ``int``, convert to int explicitly with ``~int(x)``. (Contributed by Tim Hoffmann
```

JelleZijlstra reviewed (commented):
The new behavior should also be mentioned in https://docs.python.org/3.12/library/stdtypes.html#boolean-values

It currently says "In numeric contexts (for example when used as the argument to an arithmetic operator), they behave like the integers 0 and 1, respectively.", which will no longer be true in all contexts.

timhoffm committed:
Document bool behavior for bitwise operations

This also pulls the bool type to top-level of the type
description page. Before it was only documented in the
section "Other Built-in Types / Boolean Values".

JelleZijlstra reviewed (approved):
Looks good, but I'd like to have another core dev look at the docs before we merge.

hauntsaninja commented on Doc/library/stdtypes.rst:
```diff
@@ -802,6 +802,31 @@ number, :class:`float`, or :class:`complex`::
+============================
+
+Booleans represent truth values. The :class:`bool` type has exactly two
+constant instances ``True`` and ``False``.
```
```suggestion
constant instances: ``True`` and ``False``.
```

timhoffm replied:
done.

hauntsaninja commented on Doc/library/stdtypes.rst:
```diff
@@ -802,6 +802,31 @@ number, :class:`float`, or :class:`complex`::
+Booleans represent truth values. The :class:`bool` type has exactly two
+constant instances ``True`` and ``False``.
+
+Other types can be cast to bool explicitly using the function :func:`bool`.
```
I think I preferred the old phrasing:

> The built-in function [bool()](https://docs.python.org/3/library/functions.html#bool) can be used to convert any value to a Boolean, if the value can be interpreted as a truth value (see section [Truth Value Testing](https://docs.python.org/3/library/stdtypes.html#truth) above)

It draws a clearer distinction between types and values and sidesteps the issue that `and` and `or` don't actually convert their operands to bool.

timhoffm replied:
don.

hauntsaninja commented on Doc/library/stdtypes.rst:
```diff
@@ -802,6 +802,31 @@ number, :class:`float`, or :class:`complex`::
+When applying the bitwise operators ``&``, ``|``, ``^`` to two booleans, they
+return a bool equivalent to the logical operations "and", "or", "xor". Still,
+for ``&`` and ``|``, the logical operators ``and`` and ``or`` are preferred.
+The use of the bitwise inversion operator ``~`` is deprecated and will raise
```
Maybe use the `.. deprecated::` role to mention the deprecation

timhoffm replied:
done.

hauntsaninja reviewed (commented):
I think I'd prefer to remove the `Boolean Values` section. We can preserve the `.. _bltin-boolean-values:` and `index` roles, which should hopefully preserve links.

timhoffm committed:
Move section "Boolean Values" into "Boolean Type"

hauntsaninja commented on Doc/library/stdtypes.rst:
```diff
@@ -802,6 +802,38 @@ number, :class:`float`, or :class:`complex`::
+``or`` and ``not``.
+When applying the bitwise operators ``&``, ``|``, ``^`` to two booleans, they
+return a bool equivalent to the logical operations "and", "or", "xor". Still,
+for ``&`` and ``|``, the logical operators ``and`` and ``or`` are preferred.
```
(ignorable) How do you feel about suggesting an alternative for xor as well?
```suggestion
return a bool equivalent to the logical operations "and", "or", "xor". However,
the logical operators ``and``, ``or`` and ``!=`` should be preferred
over ``&``, ``|`` and ``^``.
```

hauntsaninja commented on Doc/library/stdtypes.rst:
```diff
@@ -802,6 +802,38 @@ number, :class:`float`, or :class:`complex`::
+   raise an error in Python 3.14.
+
+:class:`bool` is a subclass of :class:`int` (see :ref:`typesnumeric`). In
+many numeric contexts, bools behave like the integers 0 and 1, respectively.
```
So that the "respectively" makes sense
```suggestion
many numeric contexts, ``False`` and ``True`` behave like the integers 0 and 1, respectively.
```

timhoffm committed:
Update Doc/library/stdtypes.rst

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

timhoffm committed:
Update Doc/library/stdtypes.rst

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

pllim commented on Doc/library/stdtypes.rst:
```diff
@@ -5394,27 +5427,6 @@ information.  There is exactly one ``NotImplemented`` object.
 It is written as ``NotImplemented``.
 
 
-.. _bltin-boolean-values:
```
Hello. The removal of this `ref` tag broke intersphinx downstream. I have not tracked down yet which middleware is calling this tag when API docstring has a "bool" mentioned. I suspect it is numpydoc. Is it not possible to reuse this tag for your new section above?

xref https://github.com/astropy/astropy/issues/15428

hauntsaninja replied:
If you open a PR, I can backport it to the 3.12 branch

pllim replied:
Sure. Please see https://github.com/python/cpython/pull/110371 . Thank you for your consideration!
